src/lem/circuit.rs

```diff
@@ -0,0 +1,853 @@
+    }
+
+    #[inline]
+    fn allocate_preimg_component_for_slot<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        slot: &Slot,
+        component_idx: usize,
+        value: F,
+    ) -> Result<AllocatedNum<F>> {
+        allocate_num(
+            cs,
+            &format!("component {component_idx} for slot {slot}"),
+            value,
+        )
+    }
+
+    fn allocate_img_for_slot<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        slot: &Slot,
+        preallocated_preimg: Vec<AllocatedNum<F>>,
+        store: &mut Store<F>,
+    ) -> Result<AllocatedNum<F>> {
+        let cs = &mut cs.namespace(|| format!("poseidon for slot {slot}"));
+        let preallocated_img = {
+            match slot.typ {
+                SlotType::Hash2 => {
+                    hash_poseidon(cs, preallocated_preimg, store.poseidon_cache.constants.c4())?
+                }
+                SlotType::Hash3 => {
+                    hash_poseidon(cs, preallocated_preimg, store.poseidon_cache.constants.c6())?
+                }
+                SlotType::Hash4 => {
+                    hash_poseidon(cs, preallocated_preimg, store.poseidon_cache.constants.c8())?
+                }
+            }
+        };
+        Ok(preallocated_img)
+    }
+
+    /// Allocates unconstrained slots
+    fn allocate_slots<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        preimgs: &[Vec<Ptr<F>>],
+        slot_type: SlotType,
+        num_slots: usize,
+        store: &mut Store<F>,
+    ) -> Result<Vec<(Vec<AllocatedNum<F>>, AllocatedNum<F>)>> {
+        assert!(
+            preimgs.len() <= num_slots,
+            "collected preimages exceeded the number of available slots"
+        );
+
+        let mut preallocations = vec![];
+
+        // First we perform the allocations for the slots containing data collected
+        // by the interpreter
+        for (slot_idx, preimg) in preimgs.iter().enumerate() {
+            let slot = Slot {
+                idx: slot_idx,
+                typ: slot_type,
+            };
+            // Allocate the preimage because the image depends on it
+            let mut preallocated_preimg = vec![];
+
+            let mut component_idx = 0;
+            for ptr in preimg.iter() {
+                let z_ptr = store.hash_ptr(ptr)?;
+
+                // allocate pointer tag
+                preallocated_preimg.push(Self::allocate_preimg_component_for_slot(
+                    cs,
+                    &slot,
+                    component_idx,
+                    z_ptr.tag.to_field(),
+                )?);
+
+                component_idx += 1;
+
+                // allocate pointer hash
+                preallocated_preimg.push(Self::allocate_preimg_component_for_slot(
+                    cs,
+                    &slot,
+                    component_idx,
+                    z_ptr.hash,
+                )?);
+
+                component_idx += 1;
```

Comment:
Aren't you simply allocating at `2 * slot_idx`, `2 * slot_idx + 1`?

Reply:
Yeah, I wanted to go slow and explicit on this one